source/Halibut.TestProxy/HttpProxyService.cs

```diff
@@ -88,70 +123,78 @@ protected override async Task ExecuteAsync(CancellationToken stoppingToken)
-                await SendConnectResponse(sourceEndpoint, connectionRequest.HttpVersion, 502, "Bad gateway", writer, cancellationToken);
-                return;
-            }
+                cancellationToken.ThrowIfCancellationRequested();
+                try
+                {
+                    logger.LogInformation("Parsing Connection Request");
+                    connectionRequest = await ParseConnectionRequest(sourceEndpoint, reader, cancellationToken);
+                }
+                catch (Exception ex)
+                {
+                    logger.LogError(ex, "An error has occurred parsing CONNECT request from {SourceEndpoint}", sourceEndpoint);
+                    await SendConnectResponse(sourceEndpoint, "1.0", 400, "Bad request", writer, cancellationToken);
+                    return;
+                }
 
-            try
-            {
-                await SendConnectResponse(sourceEndpoint, connectionRequest.HttpVersion, 200, "Connection established", writer, cancellationToken);
+                cancellationToken.ThrowIfCancellationRequested();
+                try
+                {
+                    logger.LogInformation("Creating Proxy Connection Forwarder");
+                    await CreateProxyConnectionAndForward(client, connectionRequest, cancellationToken);
+                }
+                catch (Exception ex)
+                {
+                    logger.LogError("An error has occurred establishing proxy connection: {SourceEndpoint} <-> {DestinationEndpoint}. Error: {ErrorMessage}", sourceEndpoint, connectionRequest!.Endpoint, ex.Message);
+                    await SendConnectResponse(sourceEndpoint, connectionRequest.HttpVersion, 502, "Bad gateway", writer, cancellationToken);
+                    return;
+                }
+
+                cancellationToken.ThrowIfCancellationRequested();
+                try
+                {
+                    logger.LogInformation("Sending Connect Response");
+                    await SendConnectResponse(sourceEndpoint, connectionRequest.HttpVersion, 200, "Connection established", writer, cancellationToken);
+                }
+                catch (Exception ex)
+                {
+                    logger.LogError(ex, "An error has occurred responding to CONNECT request: {SourceEndpoint} <-> {DestinationEndpoint}", sourceEndpoint, connectionRequest!.Endpoint);
+                    return;
+                }
+
+                logger.LogInformation("CONNECT request successful: {SourceEndpoint} -> {DestinationEndpoint}", sourceEndpoint, connectionRequest.Endpoint);
             }
-            catch (Exception ex)
+            finally
             {
-                logger.LogError(ex, "An error has occurred responding to CONNECT request: {SourceEndpoint} <-> {DestinationEndpoint}", sourceEndpoint, connectionRequest!.Endpoint);
-                return;
+                await reader.CompleteAsync();
```

Comment:
Does it make sense to do a when all here?

Reply:
Yes that could be good 👍